1-Draft/RFC00XX-Native-Command-Error-Handling.md

```diff
@@ -0,0 +1,307 @@
+cat: ./nofile: No such file or directory
+returns 0
+With set -o pipefail : returns non-zero (false)
+pipe statement after failure
+cat: ./nofile: No such file or directory
+returns 1
+```
+
+Get-Content ./testfile.ps1
+
+```powershell
+Write-Output "Without `$ErrorActionPreference = 'Stop' : should return zero (true)"
+Get-Content -path ./nofile | Write-Output "pipe statement after failure"
+Write-Output "returns $?"
+Write-Output "With set -o equiv. `$ErrorActionPreference = 'Stop' : should return non-zero (false)"
+$ErrorActionPreference = 'Stop'
+Get-Content -path ./nofile | Write-Output "pipe statement after failure"
+Write-Output "returns $?"
+```
+
+./testfile.ps1
+
+```output
+Without $ErrorActionPreference = 'Stop' : should return zero (true)
+Get-Content: /Users/jasonhelmick/natcmdbash/testfile.ps1:2
+Line |
+   2 |  Get-Content -path ./nofile | Write-Output "pipe statement after failu …
+     |  ~~~~~~~~~~~~~~~~~~~~~~~~~~
+     | Cannot find path '/Users/jasonhelmick/natcmdbash/nofile' because it does not exist.
+
+returns False
+With set -o equiv. $ErrorActionPreference = 'Stop' : should return non-zero (false)
+Get-Content: /Users/jasonhelmick/natcmdbash/testfile.ps1:6
+Line |
+   6 |  Get-Content -path ./nofile | Write-Output "pipe statement after failu …
+     |  ~~~~~~~~~~~~~~~~~~~~~~~~~~
+     | Cannot find path '/Users/jasonhelmick/natcmdbash/nofile' because it does not exist.
+```
+
+> [!NOTE] A common configuration command for POSIX shells `set -euo pipefail` includes the `set -u`
+> configuration which returns an error if any variable has not been previously defined. This is
+> equivalent to the existing PowerShell `Set-StrictMode` and does not need to be addressed in this
+> RFC.
+
+### $PSNativeCommandUseErrorActionPreference
+
+The `$PSNativeCommandUseErrorActionPreference` preference variable will be implemented as a boolean.
+
+- When enabled (1), `$ErrorActionPreference` will affect native commands with described behavior.
+- When disabled (0), `$ErrorActionPreference` will have no effect to the original behavior.
+- The default value is disabled (0) to maintain backward compatibility.
+- For non-zero exit codes and except for the value `Ignore`, an `ErrorRecord` will be added to
+  `$Error` that wraps the exit code and the command executed that returned the exit code.
+
+### Error object
+
+The reported error record object will be the new type: `NativeCommandException` with the following details:
+
+| Property        | Definition
+----------------  | -------------------
+| ExitCode:       |  The exit code of the failed command.
+| ErrorID:        | `"Program {0} ended with non-zero exit code {1}"`, with the command name and the exit code, from resource string `ProgramFailedToComplete`.
+| ErrorCategory:  | `ErrorCategory.NotSpecified`.
+| object:         | Exit code
+| Source:         | The full path to the application
+| ProcessInfo     | Details of failed command including path, exit code, and PID
```

Comment:
I think it would be helpful to see the arguments a a string.  Or maybe rather than command path, show the "command line".  Thoughts?

Reply:
Is the process ID of any use given that the process no longer exists?

Reply:
> I think it would be helpful to see the arguments a a string. Or maybe rather than command path, show the "command line". Thoughts?

This is not guaranteed after terminating of the process - an attempt to read a property of the Process class is likely to raise an exception. We can report reliably only current script line string.

> Is the process ID of any use given that the process no longer exists?

This can be useful as correlation data with event logs.

Reply:
After talking this over with wg-engine group we decided that process id is useful for log correlation and arguments should not be recorded because it make contain passwords / api tokens.